odoo/tools/safe_eval.py

```diff
@@ -238,131 +825,134 @@ def const_eval(expr):
+    """Test a Python expression before evaluating it
+
+    Args:
+        expr (str): A Python expression
+        mode (str): The mode of the expression (eval, exec)
+
+    Returns:
+        str | bool: An error message if the expression is not safe, False otherwise
+    """
+    try:
+        wrapped_expr = ast.unparse(CodeChecker().visit(ast.parse(expr)))
+        compile(wrapped_expr, "<sandbox>", mode)
+    except (TypeError, NameError, SyntaxError, ValueError) as err:
+        if len(err.args) >= 2 and len(err.args[1]) >= 4:
+            error = {
+                "message": err.args[0],
+                "filename": err.args[1][0],
+                "lineno": err.args[1][1],
+                "offset": err.args[1][2],
+                "error_line": err.args[1][3],
+            }
+
+            msg = f"{type(err).__name__} : {error['message']} at line {error['lineno']}\n{error['error_line']}"
+        else:
+            msg = f"{err}"
+        return msg
+    return False
+
+
+def safe_eval(
+    expr: str,
+    globals_dict: Optional[dict[str, Any]] = None,
+    locals_dict: Optional[dict[str, Any]] = None,
+    mode: str = "eval",
+    filename: Optional[str] = None,
+    sandbox_instances: tuple[type] = (),
+    sandbox_types: tuple[type] = (),
+    ast_subset: frozenset[Type[ast.AST]] | tuple[Type[ast.AST]] = _ALLOWED_NODES, # pylint: disable=bad-whitespace
+) -> Any:
+    def sanitize_context(context: dict[str, Any]) -> dict[str, Any]:
+        context = {k: v for k, v in context.items() if "__" not in k}
+        return context.copy()
+
+    if isinstance(expr, types.CodeType):
         raise TypeError("safe_eval does not allow direct evaluation of code objects.")
+    elif isinstance(expr, bytes):
+        expr = expr.decode()
 
-    # prevent altering the globals/locals from within the sandbox
-    # by taking a copy.
-    if not nocopy:
-        # isinstance() does not work below, we want *exactly* the dict class
-        if (globals_dict is not None and type(globals_dict) is not dict) \
-                or (locals_dict is not None and type(locals_dict) is not dict):
-            _logger.warning(
-                "Looks like you are trying to pass a dynamic environment, "
-                "you should probably pass nocopy=True to safe_eval().")
-        if globals_dict is not None:
-            globals_dict = dict(globals_dict)
-        if locals_dict is not None:
-            locals_dict = dict(locals_dict)
-
-    check_values(globals_dict)
-    check_values(locals_dict)
+    for node in ast_subset:
+        if node not in _ALLOWED_NODES:
+            raise ValueError(f"unsafe node: {node.__name__} (node not allowed)")
 
     if globals_dict is None:
         globals_dict = {}
+        globals_dict_eval = {}
+    else:
+        globals_dict_eval = sanitize_context(globals_dict)
+
+    if locals_dict is None:
+        locals_dict = {}
+        locals_dict_eval = {}
+    else:
+        locals_dict_eval = sanitize_context(locals_dict)
```

Comment:
with that change tests are green, can you update `test_17_globals_dict_leak_builtins` to test that line too ?
```suggestion
        locals_dict_eval = locals_dict